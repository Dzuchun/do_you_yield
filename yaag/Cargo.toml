[package]
name = "yaag"
version = "0.1.0"
edition.workspace = true

[dev-dependencies]
trybuild = "1.0"
futures-util = "0.3.31"
tokio = { version = "1.48", features = [ "full" ] }

[dependencies]
yaag_proc_macro.workspace = true
futures-core = { version = "0.3.31", optional = true }

[features]
default = [ "async" ]
async = [ "futures-core" ]
