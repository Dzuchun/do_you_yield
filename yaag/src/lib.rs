#![cfg_attr(not(test), no_std)]

pub use yaag_proc_macro::gn;

#[doc(hidden)]
pub mod sync;

#[doc(hidden)]
#[cfg(feature = "async")]
pub mod not_sync;

mod waker;
