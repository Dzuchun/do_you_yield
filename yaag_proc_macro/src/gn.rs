use crate::trans::Trans;
use syn::{Block, Expr, Ident, Token, Type, parse::Parse, parse_quote};

pub struct Input {
    is_async: bool,
    is_move: bool,
    code: Block,
    out: Type,
}

impl Parse for Input {
    fn parse(input: syn::parse::ParseStream) -> syn::Result<Self> {
        let mut is_async = false;
        let mut is_move = false;
        while !input.peek(Ident) {
            if input.parse::<Token![async]>().is_ok() {
                is_async = true;
                continue;
            }
            if input.parse::<Token![move]>().is_ok() {
                is_move = true;
                continue;
            }
            return Err(input.error("Expected move or async"));
        }
        let r#gen = input.parse::<Ident>().unwrap();
        if r#gen != "gen" {
            return Err(syn::Error::new_spanned(r#gen, "Expected `gen` keyword"));
        }
        let code = input.parse::<Block>()?;
        let _ = input.parse::<Token![->]>()?;
        let out = input.parse::<Type>()?;
        if !input.is_empty() {
            return Err(input.error("Extra input"));
        }
        Ok(Self {
            is_async,
            is_move,
            code,
            out,
        })
    }
}

impl Input {
    pub fn expand(mut self) -> Expr {
        let mv: Option<Token![move]> = self.is_move.then(Default::default);
        let out = self.out;
        self.code.trans(&out, self.is_async);
        let code = self.code;
        let module: Ident = if self.is_async {
            parse_quote!(not_sync)
        } else {
            parse_quote!(sync)
        };
        parse_quote! {{
            #[allow(unused_unsafe)]
            let fut = async #mv #code;
            ::yaag::#module::Gn {
                fut,
                _ph: ::core::marker::PhantomData::<#out>,
            }
        }}
    }
}
