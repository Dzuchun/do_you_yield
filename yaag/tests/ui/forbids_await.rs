use yaag::gn;

fn main() {
    gn! {
        gen {
            core::future::ready::<i32>(42).await;
        } -> i32
    };
}
