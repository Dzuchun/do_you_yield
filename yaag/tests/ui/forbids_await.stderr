error:
       Await outside of async context.
 --> tests/ui/forbids_await.rs:6:13
  |
6 |             core::future::ready::<i32>(42).await;
  |             ^^^^
